remote-trait-object/src/service/id.rs

```diff
@@ -0,0 +1,66 @@
+// Copyright 2020 Kodebox, Inc.
+// This file is part of CodeChain.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use super::MethodId;
+use linkme::distributed_slice;
+use serde::{Deserialize, Serialize};
+use std::collections::{HashMap, HashSet};
+
+pub const ID_ORDERING: std::sync::atomic::Ordering = std::sync::atomic::Ordering::SeqCst;
+pub type MethodIdAtomic = std::sync::atomic::AtomicU32;
+
+// linkme crate smartly collects all the registrations generated by the proc-macro
+// into a sinlge array in the link time.
+// Note that too long linkme-related variable name would cause serious compiler error in MacOS
+// So we deliberately make it have a short name
+
+// Id of methods in services.
+// Note that here the two strings mean (trait name, method name)
+// Also you can skip calling this, then the method id will be set up for default value
+// decided by the order of declaration.
+type MethodIdentifierSetter = fn(id: MethodId);
+#[distributed_slice]
+pub static MID_REG: [(&'static str, &'static str, MethodIdentifierSetter)] = [..];
+
+/// This will be provided by the coordinator.
+#[derive(PartialEq, Serialize, Deserialize, Debug, Clone)]
+pub struct IdMap {
+    // This is system-wide; All module will get same ones
+    pub method_map: Option<HashMap<(String, String), MethodId>>,
+}
+
+/// This is supposed to be called only once during the entire lifetime of module instance.
+/// However it is ok to call multiple times if the IdMap is identical, especially in the
+/// tests where each test share that static id list
+pub fn setup_identifiers(descriptor: &IdMap) {
+    // distributed_slices integrity test
+    {
+        let mut bucket: HashSet<(String, String)> = HashSet::new();
+        for (ident1, ident2, _) in MID_REG {
+            bucket.insert(((*ident1).to_owned(), (*ident2).to_owned()));
+        }
+        assert_eq!(bucket.len(), MID_REG.len());
```

Comment:
We must give a meaningful error message when the names are duplicated unexpectedly.